lib/nmatrix/shortcuts.rb

```diff
@@ -322,6 +322,76 @@ def random(shape, opts={})
 
     #
     # call-seq:
+    #     random_ortho(shape) -> NMatrix
+    #
+    # Randomly generates a +:dense+ orthogonal NMatrix, or a unitary complex NMatrix if dtype is complex. 
+    # The parameter is the dimension of the matrix. For any non-integer dtype the algorithm creates a Householder 
+    # matrix from random numbers between 0 and 1 generated by +Random::rand+. For any integer dtype an orthogonal
+    # matrix is a random column permutation of a diagonal matrix with entries equal to +1 or -1 randomly 
+    # distributed along its diagonal. You may not currently generate random numbers for a rational matrix.
+    #
+    # * *Arguments* :
+    #   - +shape+ -> Array or integer specifying the dimensions 
+    #   (note: an othogonal matrix must be square by definition).
+    # * *Returns* :
+    #   - A randomly generated orthogonal NMatrix (unitary complex for complex dtype).
+    #
+    # Examples:
+    #
+    #  m = NMatrix.rand_ortho(3) #  => -0.81460661007588 -0.05620899291406391 -0.5772838296161336
+    #                                 -0.05620899291406391 0.9982588783338108 -0.017881860733958227
+    #                                 -0.5772838296161336 -0.017881860733958227 0.8163477317420696 
+    #  m.dot m.transpose #  => 0.9999999999999991 -2.6020852139652106e-17 -2.7755575615628914e-16
+    #                          -2.6020852139652106e-17 1.0                -3.469446951953614e-18
+    #                          -2.7755575615628914e-16 -3.469446951953614e-18 1.0
+    #  m.transpose.dot m #  => 0.9999999999999991 -2.6020852139652106e-17 -2.7755575615628914e-16
+    #                         -2.6020852139652106e-17 1.0                 -3.469446951953614e-18
+    #                         -2.7755575615628914e-16 -3.469446951953614e-18 1.0
+    # References:
+    # 
+    #   Stewart 1980, "The Efficient Generation of Random Orthogonal Matrices 
+    #   with an Application to Condition Estimators"
+    #
+    def random_ortho(shape, opts={})
+      raise(NotImplementedError, "does not support rational random number generation") if [:rational64, :rational128].include? opts[:dtype]
+
+      if shape.is_a?(Fixnum)
+        n = shape 
+      elsif shape.is_a?(Array)
+        raise(ShapeError, "Expected 2D matrix") unless shape.length == 2
+        raise(ShapeError, "Expected square matrix") unless (shape[0] == shape[1])
+        n = shape[0] 
+      else
+        raise(ShapeError, "Expected 2D matrix")
+      end
+
+      # integer valued orthogonal
+      if [:byte, :int8, :int16, :int32, :int64].include? opts[:dtype]
+        d = n.times.map { [-1,1].sample }
+        q = NMatrix.diagonal(d)
+        rand_order = (0...n).to_a.sample(n)
+        q.permute_columns!(rand_order)
```

Comment:
Yes, unfortunately they don't work the same way as matlab or R; they follow the LAPACK convention of the ith element being the index of the column to swap the ith column with, having already applied all earlier swaps.  In my opinion, the matlab/R way is *way* more intuitive, but LAPACK is being used under the hood to do the permutations, so we're kind of stuck with that convention (though, we could of course write a function that translates between the two).

It appears that the reason that [2, 1, 0] works is that lapack and nmatrix ignore the last swap, since it's superfluous, so my counterexample needs to be 5x5:

Initial: (`N.diagonal([1 2 3 4 5])`)
```
[
  [1, 0, 0, 0, 0]
  [0, 2, 0, 0, 0]
  [0, 0, 3, 0, 0]
  [0, 0, 0, 4, 0]
  [0, 0, 0, 0, 5]
]
```

Desired:
```
[
  [0, 0, 1, 0, 0]
  [0, 2, 0, 0, 0]
  [3, 0, 0, 0, 0]
  [0, 0, 0, 4, 0]
  [0, 0, 0, 0, 5]
]
```
(via `permute_columns([2, 1, 2, 3, 4])`

Whereas the normal matlab/R-style permutation, `permute_columns([2, 1, 0, 3, 4])` gives:
```
[
  [1, 0, 0, 0, 0]
  [0, 2, 0, 0, 0]
  [0, 0, 3, 0, 0]
  [0, 0, 0, 4, 0]
  [0, 0, 0, 0, 5]
]
```

Now that I know the last swap is ignored, I'm not quite as sure that it's impossible to get all permutations with non-repeated indices.  Will need to think about this a bit more.